src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -0,0 +1,333 @@
+         *
+         * @return KNNEngine associated with index allocation
+         */
+        public KNNEngine getKnnEngine() {
+            return knnEngine;
+        }
+
+        /**
+         * Getter for the path to the file from which the index was loaded.
+         *
+         * @return indexPath to index
+         */
+        public String getIndexPath() {
+            return indexPath;
+        }
+
+        /**
+         * Getter for the OpenSearch index associated with the native index.
+         *
+         * @return OpenSearch index name
+         */
+        public String getOpenSearchIndexName() {
+            return openSearchIndexName;
+        }
+    }
+
+    /**
+     * Represents training data that has been allocated in native memory.
+     */
+    class TrainingDataAllocation implements NativeMemoryAllocation {
+
+        private volatile boolean closed;
+        private long pointer;
+        private final long size;
+
+        // Implement reader/writer with semaphores to deal with passing lock conditions between threads
+        private int readCount;
+        private Semaphore readSemaphore;
+        private Semaphore writeSemaphore;
+
+        /**
+         * Constructor
+         *
+         * @param pointer pointer in memory to the training data allocation
+         * @param size amount memory needed for allocation in kilobytes
+         */
+        TrainingDataAllocation(long pointer, long size) {
+            this.closed = false;
+            this.pointer = pointer;
+            this.size = size;
+
+            this.readCount = 0;
+            this.readSemaphore = new Semaphore(1);
+            this.writeSemaphore = new Semaphore(1);
+        }
+
+        @Override
+        public void close() {
+            if (closed) {
+                return;
+            }
+
+            closed = true;
+            if (this.pointer != 0) {
+                JNIService.freeVectors(this.pointer);
+            }
+        }
+
+        @Override
+        public boolean isClosed() {
+            return closed;
+        }
+
+        @Override
+        public long getPointer() {
+            return pointer;
+        }
+
+        /**
+         * A read lock will be obtained when a training job needs access to the TrainingDataAllocation.
+         * In the future, we may want to switch to tryAcquire functionality.
+         */
+        @Override
+        public void readLock() {
+            try {
+                readSemaphore.acquire();
+            } catch (InterruptedException ex) {
+                throw new RuntimeException(ex);
+            }
+
+
+            if (readCount == 0) {
+                try {
+                    writeSemaphore.acquire();
+                } catch (InterruptedException e) {
+                    readSemaphore.release();
+                    throw new RuntimeException(e);
+                }
+            }
+
+            readCount++;
+            readSemaphore.release();
```

Comment:
So, I dont think so. If we always release the readSemaphore at the end, but the readSemaphore was not acquired, it will add an additional permit to the readSemaphore. In other words, you do not need to acquire a semaphore to call release on it. 

I could add incrementing readCount into the try block with the writelock, but I was thinking it would be better to keep the try/catch simple and focused on locking.